### PR TITLE
ASoC: Intel: add rt712 support

### PR DESCRIPTION
**Pull request by bardliao**
Rt712 is a multi function codec which shpports headset, amp, and dmic functions. Rt712 has two sdw interfaces and codec drivers, one for jack and amp, the other for dmic. part id 0x712 is for jack and amp, and 0x1712 is for dmic.

Topology change: https://github.com/thesofproject/sof/pull/7634

### CODE REVIEW

File: sound/soc/intel/boards/sof_sdw.c [hunk 1]
```diff
@@ -599,6 +599,42 @@ static struct sof_sdw_codec_info codec_info_list[] = {
...
+			{
+				.direction = {false, true},
+				.dai_name = "rt712-sdca-dmic-aif1",
+				.dai_type = SOF_SDW_DAI_TYPE_MIC,
+				.dailink = {SDW_UNUSED_DAI_ID, SDW_DMIC_DAI_ID},
+				.init = sof_sdw_rt712_sdca_dmic_init,
+			},
+		},
+		.dai_num = 1,
+	},
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c:652]: move this block after line 652 to keep all rt711 together?

File: sound/soc/intel/boards/sof_sdw.c [hunk 2]
```diff
@@ -599,6 +599,42 @@ static struct sof_sdw_codec_info codec_info_list[] = {
...
+				.dai_type = SOF_SDW_DAI_TYPE_JACK,
+				.dailink = {SDW_JACK_OUT_DAI_ID, SDW_JACK_IN_DAI_ID},
+				.init = sof_sdw_rt712_sdca_init,
+				.exit = sof_sdw_rt712_sdca_exit,
+			},
+			{
+				.direction = {true, false},
+				.dai_name = "rt712-sdca-aif2",
+				.dai_type = SOF_SDW_DAI_TYPE_AMP,
+				.dailink = {SDW_AMP_OUT_DAI_ID, SDW_AMP_IN_DAI_ID},
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: .dailink = {SDW_AMP_OUT_DAI_ID, SDW_UNUSED_DAI_ID} ?

if the direction for capture is false, why would we have an AMP_IN_DAI_ID?

File: sound/soc/intel/boards/sof_sdw_rt712_sdca.c [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+ * Note this MUST be called before snd_soc_register_card(), so that the props
+ * are in place before the codec component driver's probe function parses them.
+ */
+static int rt712_sdca_add_codec_device_props(struct device *sdw_dev)
+{
+	struct property_entry props[MAX_NO_PROPS] = {};
+	struct fwnode_handle *fwnode;
+	int ret;
+
+	if (!SOF_RT711_JDSRC(sof_sdw_quirk))
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw_rt712_sdca.c]: RT711?

File: sound/soc/intel/boards/sof_sdw_rt712_sdca.c [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+	fwnode = fwnode_create_software_node(props, NULL);
+	if (IS_ERR(fwnode))
+		return PTR_ERR(fwnode);
+
+	ret = device_add_software_node(sdw_dev, to_software_node(fwnode));
+
+	fwnode_handle_put(fwnode);
+
+	return ret;
+}
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw_rt712_sdca.c]: this entire function looks like a helper that could be reused between RT711 and RT712?

File: sound/soc/intel/boards/sof_sdw_rt712_sdca.c [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+
+int sof_sdw_rt712_sdca_exit(struct snd_soc_card *card, struct snd_soc_dai_link *dai_link)
+{
+	struct mc_private *ctx = snd_soc_card_get_drvdata(card);
+
+	if (!ctx->headset_codec_dev)
+		return 0;
+
+	if (!SOF_RT711_JDSRC(sof_sdw_quirk))
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw_rt712_sdca.c]: rt711?

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c [hunk 6]
```diff
@@ -170,6 +170,24 @@ static const struct snd_soc_acpi_adr_device rt1316_1_single_adr[] = {
 	}
 };
 
+static const struct snd_soc_acpi_adr_device rt712_0_single_adr[] = {
+	{
+		.adr = 0x000030025D071201ull,
+		.num_endpoints = 1,
+		.endpoints = &single_endpoint,
```
Comment (plbossart) [inline sound/soc/intel/common/soc-acpi-intel-tgl-match.c]: we're not using the endpoints for now, but I don't think this is correct. We have two with headphone/jack and amplifier, no?

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c [hunk 7]
```diff
@@ -435,6 +467,12 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_tgl_sdw_machines[] = {
 		.drv_name = "sof_sdw",
 		.sof_tplg_filename = "sof-tgl-rt715-rt711-rt1308-mono.tplg",
 	},
+	{
+		.link_mask = 0xF, /* 4 active links required */
+		.links = tgl_712_only,
+		.drv_name = "sof_sdw",
+		.sof_tplg_filename = "sof-tgl-rt712.tplg",
+	},
```
Comment (plbossart) [inline sound/soc/intel/common/soc-acpi-intel-tgl-match.c:490]: should we add this for ADL/RPL/MTL as well?

**Review by plbossart — CHANGES_REQUESTED** [review]
that's a good start @bardliao, couple of minor comments below.

File: sound/soc/intel/boards/sof_sdw.c (hunk 1 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:652]: > move this block after line 652 to keep all rt711 together?

ack

File: sound/soc/intel/boards/sof_sdw.c (hunk 2 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > .dailink = {SDW_AMP_OUT_DAI_ID, SDW_UNUSED_DAI_ID} ?
> 
> if the direction for capture is false, why would we have an AMP_IN_DAI_ID?

Right, replaced with SDW_UNUSED_DAI_ID

File: sound/soc/intel/boards/sof_sdw_rt712_sdca.c (hunk 3 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw_rt712_sdca.c]: > RT711?

Hmm, I don't want to waste quirk bits, and SOF_RT711_JDSRC can apply to all Realtek jacks. I will rename it to `SOF_JACK_JDSRC`

File: sound/soc/intel/boards/sof_sdw_rt712_sdca.c (hunk 4 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw_rt712_sdca.c]: > this entire function looks like a helper that could be reused between RT711 and RT712?

Indeed, I will reuse those helper functions.

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c (hunk 7 above)
Comment (bardliao) [reply sound/soc/intel/common/soc-acpi-intel-tgl-match.c:490]: > should we add this for ADL/RPL/MTL as well?

That is an open. The target is MTL, but currently, it is only tested on TGL Up Xtreme board. I don't know is ADL/RPL needed.

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c (hunk 6 above)
Comment (bardliao) [reply sound/soc/intel/common/soc-acpi-intel-tgl-match.c]: > we're not using the endpoints for now, but I don't think this is correct. We have two with headphone/jack and amplifier, no?

Added rt712_endpoints[] for rt712.

File: sound/soc/intel/boards/sof_sdw_rt_sdca_jack_common.c [hunk 8]
```diff
@@ -17,11 +17,13 @@
 #include <sound/jack.h>
 #include "sof_sdw_common.h"
 
+static int part_id;
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw_rt_sdca_jack_common.c]: do we really want a static variable? IIRC it's possible to have solutions with multiple jacks for e.g. front and back

File: sound/soc/intel/boards/sof_sdw_rt_sdca_jack_common.c [hunk 9]
```diff
@@ -152,6 +152,9 @@ int sof_sdw_rt_sdca_jack_exit(struct snd_soc_card *card, struct snd_soc_dai_link
 	if (!ctx->headset_codec_dev)
 		return 0;
 
+	if (!SOF_JACK_JDSRC(sof_sdw_quirk))
+		return 0;
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw_rt_sdca_jack_common.c:159]: oh, isn't that a bug you want to correct before moving the code to a different file? That would make backports easier if indeed this is a problem

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c [hunk 10]
```diff
@@ -435,6 +482,12 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_tgl_sdw_machines[] = {
 		.drv_name = "sof_sdw",
 		.sof_tplg_filename = "sof-tgl-rt715-rt711-rt1308-mono.tplg",
 	},
+	{
+		.link_mask = 0xF, /* 4 active links required */
```
Comment (plbossart) [inline sound/soc/intel/common/soc-acpi-intel-tgl-match.c:486]: can you remind me why we need the 4 links? This is UpExtreme so presumably you are using link0 only, no?

**Review by plbossart — COMMENTED** [review]
one big doubt on the 'part_id' static variable @bardliao, plus a couple of comments below.

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c (hunk 10 above)
Comment (bardliao) [reply sound/soc/intel/common/soc-acpi-intel-tgl-match.c:486]: > can you remind me why we need the 4 links? This is UpExtreme so presumably you are using link0 only, no?

I use UpExtreme  indeed, and enable 4 links via selecting AAM to SoundWire.

File: sound/soc/intel/boards/sof_sdw_rt_sdca_jack_common.c (hunk 8 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw_rt_sdca_jack_common.c]: > do we really want a static variable? IIRC it's possible to have solutions with multiple jacks for e.g. front and back

@plbossart Thanks for the reminding. Using component->name_prefix to determine codecs now.

File: sound/soc/intel/boards/sof_sdw_rt_sdca_jack_common.c (hunk 9 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw_rt_sdca_jack_common.c:159]: > oh, isn't that a bug you want to correct before moving the code to a different file? That would make backports easier if indeed this is a problem

ack

**Review by plbossart — APPROVED** [review]
LGTM @bardliao, thanks!

**Review by shumingfan — APPROVED** [review]
LGTM. Thanks. @bardliao

**Pull request merged**